packages/nft/src/generators/artZero.ts

```
@@ -0,0 +1,137 @@
+  /** e.g. ['1713843826001', '1745379826001'], [] */
+  attributesValue: string[]
+  /** e.g. 0 */
+  listed_date: number
+  /** e.g. 0 */
+  price: number
+  /** e.g. false */
+  is_for_sale: boolean
+  /** e.g. '' */
+  nft_owner: string
+  /** e.g. null */
+  highest_bid: unknown
+  /** e.g. false */
+  is_locked: boolean
+  /** e.g. { 'Registration Time': '1713843826001', 'Expiration Time': '1745379826001' }, { Background: 'Happy Dev Blue', Character: 'Adam Gagol' } */
+  traits: Record<string, string | undefined>
+  /** e.g. null */
+  hash: unknown
+  /** e.g. null */
+  flag: unknown
+  /** e.g. null */
+  dataObject: unknown
+  /** e.g. '2024-04-23T03:43:56.671Z' */
+  createdTime: string
+  /** e.g. '2024-04-23T03:44:08.273Z' */
+  updatedTime: string
+}>
+
+export const createArtZeroNftAsyncGenerator: CreateNftAsyncGenerator<Nft<'artzero', 'aleph-zero'>> = async function* (
+  address,
+  { batchSize }
+) {
+  let offset = 0
+
+  while (true) {
+    const owner = encodeAddress(address, 42)
+    const limit = batchSize
+    const sort = -1
+    const isActive = true
+
+    const response = await fetch(`${ARTZERO_API}/getNFTsByOwner`, {
+      method: 'POST',
+      headers: {
+        Accept: '*/*',
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: new URLSearchParams(
+        Object.entries({ owner, limit, offset, sort, isActive }).map(([key, value]) => [key, value.toString()])
+      ).toString(),
+    })
+    if (!response.ok) throw new Error(`ArtZero api response not ok: ${response.status}`)
+
+    const result: ArtzeroGetNFTsByOwnerResult = await response.json()
+    if (!Array.isArray(result?.ret)) break
+    if ((result?.ret?.length ?? 0) < 1) break
+
+    const type = 'artzero' as const
+    const chain = 'aleph-zero' as const
+
+    yield* (result?.ret ?? []).map(
+      (nft, index): Nft<'artzero', 'aleph-zero'> => ({
+        type,
+        chain,
+        id: `${type}-${chain}-${nft._id ?? index + offset}`,
```

Is the `_id` field optional?

---

I don’t trust the API

---

In that case, I think it's better to just throw an error rather than returning an item without it actual ID. Because the ID will potentially be stored in local storage if the user favourite or hide the NFT.